File: testing/testsrc/nn.py
import numpy as np
from .activation import *

class nn:
  def __init__(self, inshape, outshape, activation, learningrate, weightsinit=(-1, 1), biasesinit=(0, 0)):
    self.weightsinit, self.biasesinit = weightsinit, biasesinit
    self.inshape, self.outshape = inshape, outshape
    self.activation, self.learningrate = activation, learningrate
    
    self.weights = np.random.uniform(weightsinit[0], weightsinit[1], (np.prod(inshape), np.prod(outshape)))
    self.biases = np.random.uniform(biasesinit[0], biasesinit[1], np.prod(outshape))
    
  def __copy__(self):
    return type(self)(self.inshape, self.outshape, self.activation, self.learningrate, self.weightsinit, self.biasesinit)

  def forward(self, input):
    self.input = input.flatten()
    self.out = self.weights.T @ self.input + self.biases
    
    self.out = self.activation.forward(self.out)

    return self.out.reshape(self.outshape)

  def backward(self, outError):
    outError = outError.flatten()
    
    outGradient = self.activation.backward(self.out) * outError
      
    inputError = self.weights @ outError
      
    self.biases += outGradient * self.learningrate
    self.weights += np.outer(self.input.T, outGradient) * self.learningrate
    return inputError.reshape(self.inshape)

File: testing/cppsrc/nn.cpp
#include <iostream>
#include <vector>
#include <algorithm>

#include "print.cpp"

class nn {
  public:
    vector <int> _inshape;
    vector <int> _outshape;
    int _learningrate;
    vector <int> _weightsinit;
    vector <int> _biasesinit;

    nn(vector <int> inshape, vector <int> outshape, int learningrate, vector <int> weightsinit = { -1, 1 }, vector <int> biasesinit = { 0, 0 }) {
      _inshape = inshape;
      _outshape = outshape;
      _learningrate = learningrate;
      _weightsinit = weightsinit;
      _biasesinit = biasesinit;
      
      vector <vector <double>> weights = 


    }
/*
    vector <double> forward(vector <double> input) {
      transform(input.begin(), input.end(), input.begin(), [&](double input){ return (a * input < input) ? input : 0; });
      return input;
    }

    vector <double> backward(vector <double> input) {
      transform(input.begin(), input.end(), input.begin(), [&](double input){ return (input <= 0) ? a : 1; });
      return input;
    }
*/
};

int main() {
  nn model({ 1, 6, 82 }, { 53, 6, 245 });
}
